api/src/game_state_server.py

```diff
@@ -1,108 +1,173 @@
-    start_x: int
-    start_y: int
-    start_z: int
-    end_x: int
-    end_y: int
-    end_z: int
+class MessageContents:
+    type: str
+    data: Union[str, Iterable[Union[Ping, Token]]]
+    request_id: Optional[str] = None
 
 
 @dataclass
-class Reply:
-    type: str
-    data: Union[list, str]
+class Message:
+    targets: Iterable[Hashable]
+    contents: MessageContents
 
 
 class RoomData:
     def __init__(self, room_id: str, initial_connection=None):
-        self.room_id = room_id
-        self.game_state = {}
+        self.room_id: str = room_id
+        self.game_state: Dict[str, Union[Ping, Token]] = {}
         self.id_to_positions = {}
         self.positions_to_ids = {}
         self.clients = set()
         if initial_connection:
             self.clients.add(initial_connection)
 
 
-class MessageError(Exception):
-    def __init__(self, message):
-        self.message = message
-
-
 class GameStateServer:
     def __init__(self, room_store: RoomStore):
-        self._rooms = {}
+        self._rooms: Dict[str, RoomData] = {}
         self.room_store = room_store
 
-    def new_connection_request(self, client: Hashable, room_id: str) -> Reply:
+    def new_connection_request(self, client_id: Hashable, room_id: str) -> Message:
         if self._rooms.get(room_id, False):
-            self._rooms[room_id].clients.add(client)
+            self._rooms[room_id].clients.add(client_id)
         else:
-            self._rooms[room_id] = RoomData(room_id, initial_connection=client)
+            self._rooms[room_id] = RoomData(room_id, initial_connection=client_id)
             if self.room_store.room_data_exists(room_id):
-                state_to_load = self.room_store.read_room_data(room_id)
-                for token_d in state_to_load.values():
-                    self._create_or_update_token(self._dict_to_token(token_d), room_id)
-        return Reply('state', self.get_state(room_id))
+                tokens_to_load = self.room_store.read_room_data(room_id)
+                for token_data in tokens_to_load:
+                    try:
+                        token = Token(**token_data)
+                    except TypeError:
+                        raise
+                    else:
+                        self._create_or_update_token(token, room_id)
+        return Message(
+            [client_id], MessageContents('connected', self.get_state(room_id))
+        )
 
-    def connection_dropped(self, client: Hashable, room_id: str) -> None:
+    def connection_dropped(self, client_id: Hashable, room_id: str) -> None:
         if self._rooms.get(room_id, False):
-            self._rooms[room_id].clients.remove(client)
+            self._rooms[room_id].clients.remove(client_id)
             # Save the room if the last client leaves and there is something to save
             if not self._rooms[room_id].clients and self._rooms[room_id].game_state:
                 print('Writing room data')
+                data_to_store = []
```

Comment:
this collection can be written with python's [filter](https://docs.python.org/3/library/functions.html#filter) or with a list comprehension

Reply:
But that feels like a stylistic decision

Reply:
your call
=